### ISSUE TITLE
Some questions on your "https://dustinspecker.com/posts/how-do-kubernetes-and-docker-create-ip-addresses/"

### ISSUE THREAD

**Issue by hvina** [dustinspecker/dustinspecker.com#10]
This is with reference to you [article on veth / namespaces](https://dustinspecker.com/posts/how-do-kubernetes-and-docker-create-ip-addresses/). 
**TL;DR**
When I create 2 veth pairs and associate them with 2 name spaces and I add a iptable rule for just 1 of the veth interfaces the other veth interface also is pingable from outside . 

I differed from the code and use 2 IP subnets to get over the routing problem. 
```
# Create 2 namespaces (netns_dustin, netns_leah).
# Create 2 pair of veth interfaces .Assign one of them to each namespace
# Assign ip address 10.0.0.10 and 10.0.0.11 to one pair and 10.0.1.10 10.0.1.11 to the other pair
# Add default routes in each namespace with the 10.0.0.10 and 10.0.1.10
```
192.168.1.1 is my router and enp0s25 is the interface name 
 Behavior :
If the iptables rules are not turned on then the pinging the external router from both the name spaces **do not work as expected**
```
# sudo ip netns exec netns_dustin ping 192.168.1.1
# sudo ip netns exec netns_dustin ping 8.8.8.8
# sudo ip netns exec netns_leah ping 192.168.1.1
# sudo ip netns exec netns_leah ping 8.8.8.8
```

**Turn on ip forwarding for only the veth_dustin interface** 
```
#sudo iptables --append FORWARD --in-interface veth_dustin --out-interface enp0s25 --jump ACCEPT
#sudo iptables --append FORWARD --in-interface enp0s25 --out-interface veth_dustin --jump ACCEPT
#sudo iptables --append POSTROUTING --table nat --out-interface enp0s25 --jump MASQUERADE

# sudo ip netns exec netns_dustin ping 192.168.1.1
# sudo ip netns exec netns_dustin ping 8.8.8.8
# sudo ip netns exec netns_leah ping 192.168.1.1
# sudo ip netns exec netns_leah ping 8.8.8.8

all of them work. 
```
This is the output of the iptables-save 
```
# Generated by iptables-save v1.8.4 on Sat Dec  3 13:11:29 2022
*nat
:PREROUTING ACCEPT [13:723]
:INPUT ACCEPT [11:555]
:OUTPUT ACCEPT [7:381]
:POSTROUTING ACCEPT [2:146]
-A POSTROUTING -o enp0s25 -j MASQUERADE
COMMIT
# Completed on Sat Dec  3 13:11:29 2022
# Generated by iptables-save v1.8.4 on Sat Dec  3 13:11:29 2022
*filter
:INPUT ACCEPT [179:12408]
:FORWARD ACCEPT [12:1008]
:OUTPUT ACCEPT [130:18885]
-A FORWARD -i veth_dustin -o enp0s25 -j ACCEPT
-A FORWARD -i enp0s25 -o veth_dustin -j ACCEPT
COMMIT
# Completed on Sat Dec  3 13:11:29 2022
```

**Question**
 if the iptables are turned on only for ns_dustin then pinging of ns_leah also works . why is that ? how can the ipforwarding rule for veth_dustin forward packets to veth_leah . Is this a bug in Linux ? Should the iptable forwarding rules be written differently ?

**Comment by hvina** [comment]
This is the set of commands I tried 
```
sudo ip netns add netns_dustin
sudo ip netns exec netns_dustin ip link set dev lo up
sudo ip link add dev veth_dustin type veth peer name veth_ns_dustin
sudo ip link set dev veth_dustin up
sudo ip link set veth_ns_dustin netns netns_dustin
sudo ip netns exec netns_dustin ip link set dev veth_ns_dustin up
sudo ip address add 10.0.0.10/24 dev veth_dustin
sudo ip netns exec netns_dustin ip address add 10.0.0.11/24 dev veth_ns_dustin
sudo ip netns exec netns_dustin ip route add default via 10.0.0.10
echo 1 | sudo tee /proc/sys/net/ipv4/ip_forward
echo "nameserver 192.168.1.1" | sudo tee /etc/netns/netns_dustin/resolv.conf
```
For the other namespace 
```
sudo ip link add dev veth_leah type veth peer name veth_ns_leah
sudo ip link set dev veth_leah up
sudo ip address add 10.0.1.10/24 dev veth_leah
sudo ip netns add netns_leah
sudo ip link set dev veth_ns_leah netns netns_leah
sudo ip netns exec netns_leah ip link set dev lo up
sudo ip netns exec netns_leah ip link set dev veth_ns_leah up
sudo ip netns exec netns_leah ip address add 10.0.1.11/24 dev veth_ns_leah
sudo ip netns exec netns_leah ip route add default via 10.0.1.10

sudo mkdir -p /etc/netns/netns_leah
echo "nameserver 192.168.1.1" | sudo tee /etc/netns/netns_leah/resolv.conf
```
And these commands enabled the iptable forwarding and nat
```
sudo iptables --append FORWARD --in-interface veth_dustin --out-interface enp0s25 --jump ACCEPT
sudo iptables --append FORWARD --in-interface enp0s25 --out-interface veth_dustin --jump ACCEPT
sudo iptables --append POSTROUTING --table nat --out-interface enp0s25 --jump MASQUERADE
```
